natlib/svcd.c

```diff
@@ -183,5 +192,39 @@ static int svcd_init( lua_State *L )
         lua_call(L, 3, 0);
     }
 
+    
+    return 0;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// SVCD.ndispatch((string pay, number srcip, number srcport)) 
+// Authors: Jennifer Dai, Prashan Dharmasena, Wenqin 
+/////////////////////////////////////////////////////////////
+
+static int svcd_ndispatch( lua_State *L )
+{
+    // Checks to make sure it is passed 3 parameters
+    if (lua_gettop(L) != 3) return luaL_error(L, "Expected (pay, srcip, srcport)"); 
+
+    size_t parlen;
+    const char* pay = lua_tolstring(L, -1, &parlen); 
+    uint16_t ivkid = lua_tonumber(L, -1); // changes pay parameter to number and returns
+    lua_getglobal(L, "SVCD"); // gets table
+    lua_pushstring(L, "oursubs"); // use oursubs as the key
+    lua_gettable(L, 4);
+    char ivkidstr[16];
+    sprintf(ivkidstr, "%d", ivkid);
+    lua_pushstring(L, ivkidstr);
+    lua_gettable(L, 5);
+
+    size_t size;
+    const char* item = lua_tolstring(L, 6, &size);
+    if (!lua_isnil(L, 5)) {
+	char newstr[3];
+	strncpy(newstr, pay + 3, parlen - 3); // gets from 3rd character to end of string
```

Comment:
WAT! strncpy _always_ takes the array size, (or size-1 if you want to make sure it is null terminated). You are creating an array of size three, and then writing parnel-3 bytes to it, almost _certainly_ more than three bytes!

Besides, why copy the string at all? you can just call lua_pushlstring(L, pay+3, parlen-3); 

Not also that you cannot use pushstring - this is not a CStr, so how would pushstring know the length?